extensions/parquet/table/src/main/java/io/deephaven/parquet/table/util/S3SeekableByteChannel.java

```diff
@@ -0,0 +1,320 @@
+        return Math.toIntExact(byteNumber / fragmentSize);
+    }
+
+    private CompletableFuture<ByteBuffer> computeFragmentFuture(final int fragmentIndex) {
+        final long readFrom = (long) fragmentIndex * fragmentSize;
+        final long readTo = Math.min(readFrom + fragmentSize, size) - 1;
+        final String range = "bytes=" + readFrom + "-" + readTo;
+        return s3AsyncClient.getObject(builder -> builder.bucket(bucket).key(key).range(range),
+                new ByteBufferAsyncResponseTransformer<>(fragmentSize));
+    }
+
+    private void handleS3Exception(final Exception e, final String operationDescription) {
+        if (e instanceof InterruptedException) {
+            Thread.currentThread().interrupt();
+            throw new CancellationException("Thread interrupted while " + operationDescription, e);
+        } else if (e instanceof ExecutionException) {
+            throw new UncheckedDeephavenException("Execution exception occurred while " + operationDescription, e);
+        } else if (e instanceof TimeoutException) {
+            throw new UncheckedDeephavenException("Operation timeout while " + operationDescription + " after waiting " +
+                    "for duration " + readTimeout, e);
+        }
+    }
+
+    @Override
+    public int write(final ByteBuffer src) throws IOException {
+        throw new NonWritableChannelException();
+    }
+
+    @Override
+    public long position() throws ClosedChannelException {
+        final long localPosition = position;
+        checkClosed(localPosition);
+        return localPosition;
+    }
+
+    @Override
+    public SeekableByteChannel position(final long newPosition) throws ClosedChannelException {
+        if (newPosition < 0) {
+            throw new IllegalArgumentException("newPosition cannot be < 0, provided newPosition=" + newPosition);
+        }
+        checkClosed(position);
+        position = newPosition;
+        return this;
+    }
+
+    @Override
+    public long size() throws ClosedChannelException {
+        checkClosed(position);
+        if (size < 0){
+            populateSize();
+        }
+        return size;
+    }
+
+    private void populateSize() {
+        if (context.getSize() < 0) {
+            // Fetch the size of the file on the first read using a blocking HEAD request
+            final HeadObjectResponse headObjectResponse;
+            try {
+                headObjectResponse = s3AsyncClient.headObject(builder -> builder.bucket(bucket).key(key))
+                        .get(readTimeout.toNanos(), TimeUnit.NANOSECONDS);
+            }  catch (final InterruptedException | ExecutionException | TimeoutException e) {
+                final String operation = "fetching HEAD for file " + key + " in S3 bucket " + bucket;
+                handleS3Exception(e, operation);
+                throw new UncheckedDeephavenException("Exception caught while " + operation, e);
```

Comment:
This line will never be executed. Maybe instead have it return exception and do `throw handleS3Exception(...)`?